tcp_packet_conn.go

```diff
@@ -103,14 +103,13 @@ func (t *tcpPacketConn) startReading(conn net.Conn) {
 
 func (t *tcpPacketConn) handleRecv(pkt streamingPacket) {
 	t.mu.Lock()
+	defer t.mu.Unlock()
```

Comment:
I think `close(t.closedChan)` should be out of the lock on `Close()` to avoid deadlock if `handleRecv` takes lock during below select.
https://github.com/pion/ice/blob/8f4d021fc7e54cbeb9ac5633c80ef19b92e1a276/tcp_packet_conn.go#L195-L201

IMO, it would be better to close `recvChan` when leaving `startReading()` instead of closing it in `Close()`, and keep unlocking before select.

Reply:
@at-wat Thanks for your review! Just trying to see if we're on the same page: am I correct when I say that the deadlock you're talking about can occur when the following conditions are met:

1. goroutine A receives a packet, it enters `handleRecv`, but the `recvChan` is full, so it's stuck at `select`, and
2. goroutine B calls `Close` but is stuck because the mutex is still acquired in `handleRecv`?

> IMO, it would be better to close recvChan when leaving startReading() instead of closing it in Close(), and keep unlocking before select.

I'm not sure that's possible in this case: multiple connections can be added in `AddConn`, and it is not known when the last one was received. All of their packets are being written to the same receiving channel. If the `recvChan` was closed as soon as one `startReading()` ends, it would cause panics on `write to close channel` for other active connections, or the ones that are yet to be added.

Reply:
I guess we could check for the "closed" state in `AddConn` and prevent adding new connections after `Close` is called, and then close the `recvChan` as soon as all created reading goroutines have ended.